libyul/backends/evm/ControlFlowGraphBuilder.cpp

```diff
@@ -0,0 +1,484 @@
+using namespace solidity::yul;
+using namespace std;
+
+std::unique_ptr<CFG> ControlFlowGraphBuilder::build(
+	AsmAnalysisInfo& _analysisInfo,
+	Dialect const& _dialect,
+	Block const& _block
+)
+{
+	auto result = std::make_unique<CFG>();
+	result->entry = &result->makeBlock();
+
+	ControlFlowGraphBuilder builder(*result, _analysisInfo, _dialect);
+	builder.m_currentBlock = result->entry;
+	builder(_block);
+
+	// Determine which blocks are reachable from the entry.
+	util::BreadthFirstSearch<CFG::BasicBlock*> reachabilityCheck{{result->entry}};
+	ranges::actions::push_back(
+		reachabilityCheck.verticesToTraverse,
+		result->functionInfo | ranges::views::values | ranges::views::transform(
+			[](auto&& _function) { return _function.entry; }
+		)
+	);
+	reachabilityCheck.run([&](CFG::BasicBlock* _node, auto&& _addChild) {
+		visit(util::GenericVisitor{
+			[&](CFG::BasicBlock::Jump& _jump) {
+				_addChild(_jump.target);
+			},
+			[&](CFG::BasicBlock::ConditionalJump& _jump) {
+				_addChild(_jump.zero);
+				_addChild(_jump.nonZero);
+			},
+			[](CFG::BasicBlock::FunctionReturn&) {},
+			[](CFG::BasicBlock::Terminated&) {},
+			[](CFG::BasicBlock::MainExit&) {}
+		}, _node->exit);
+	});
+
+	// Remove all entries from unreachable nodes from the graph.
+	for (auto* node: reachabilityCheck.visited)
+		cxx20::erase_if(node->entries, [&](CFG::BasicBlock* entry) -> bool {
+			return !reachabilityCheck.visited.count(entry);
+		});
+
+	// TODO: It might be worthwhile to run some further simplifications on the graph itself here.
+	// E.g. if there is a jump to a node that has the jumping node as its only entry, the nodes can be fused, etc.
+
+	return result;
+}
+
+ControlFlowGraphBuilder::ControlFlowGraphBuilder(
+	CFG& _graph,
+	AsmAnalysisInfo& _analysisInfo,
+	Dialect const& _dialect
+):
+m_graph(_graph),
+m_info(_analysisInfo),
+m_dialect(_dialect)
+{
+}
+
+StackSlot ControlFlowGraphBuilder::operator()(Literal const& _literal)
+{
+	return LiteralSlot{valueOfLiteral(_literal), _literal.debugData};
+}
+
+StackSlot ControlFlowGraphBuilder::operator()(Identifier const& _identifier)
+{
+	return VariableSlot{lookupVariable(_identifier.name), _identifier.debugData};
+}
+
+StackSlot ControlFlowGraphBuilder::operator()(Expression const& _expression)
+{
+	return std::visit(*this, _expression);
+}
+
+CFG::Operation& ControlFlowGraphBuilder::visitFunctionCall(FunctionCall const& _call)
+{
+	yulAssert(m_scope, "");
+	yulAssert(m_currentBlock, "");
+
+	if (BuiltinFunction const* builtin = m_dialect.builtin(_call.functionName.name))
+	{
+		CFG::Operation& operation = m_currentBlock->operations.emplace_back(CFG::Operation{
+			// input
+			_call.arguments |
+			ranges::views::enumerate |
+			ranges::views::reverse |
+			ranges::views::filter(util::mapTuple([&](size_t idx, auto const&) {
+				return !builtin->literalArgument(idx).has_value();
+			})) |
+			ranges::views::values |
+			ranges::views::transform(std::ref(*this)) |
+			ranges::to<Stack>,
```

Comment:
lol

Reply:
I don't really understand what this sequence does. Can you maybe add a short comment?

Reply:
Does this look more understandable?

```cpp
vector<stackslot> s;
for (auto [idx, arg]: _call.arguments | enumerate | reverse)
    if (builtin->literalArgument(idx).has_value())
        s.emplace_back(std::visit(*this, arg));
```

Basically just visit the arguments in reverse. Except for arguments that are not literals? For example, `loadimmutable`.

Reply:
Ah ok, yea. But I think we can keep it the current way

Reply:
I'm actually not sure if we should not replace it :)
The `transform(std::ref(*this))` call is very hidden.
Is it actually clear that these calls are done in order?

Reply:
Sure, the calls are done in order - there's no magic happening here, it's still basically range based loops, I'm just avoiding unnecessary variables :-). Also for the purpose of generating the graph the order of calls does not even matter. And not sure in what way ``transform(std::ref(*this))`` is "hidden" - it's an explicit transformation on the range right spelled out right there :-).
That being said, I'll just rewrite it :-).